File: app/src/main/java/net/tnose/app/trisquel/FilmRollFragment.java
package net.tnose.app.trisquel;

import android.content.Context;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.LinearLayoutManager;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.ArrayList;

/**
 * A fragment representing a list of Items.
 * <p/>
 * Activities containing this fragment MUST implement the {@link OnListFragmentInteractionListener}
 * interface.
 */
public class FilmRollFragment extends Fragment {

    // TODO: Customize parameter argument names
    private static final String ARG_COLUMN_COUNT = "column-count";
    // TODO: Customize parameters
    private int mColumnCount = 1;
    private OnListFragmentInteractionListener mListener;

    private RecyclerViewEmptySupport mRecyclerView;
    private ArrayList<FilmRoll> list;
    private MyFilmRollRecyclerViewAdapter filmrollRecyclerViewAdapter;
    /**
     * Mandatory empty constructor for the fragment manager to instantiate the
     * fragment (e.g. upon screen orientation changes).
     */
    public FilmRollFragment() {
    }

    // TODO: Customize parameter initialization
    @SuppressWarnings("unused")
    public static FilmRollFragment newInstance(int columnCount) {
        FilmRollFragment fragment = new FilmRollFragment();
        Bundle args = new Bundle();
        args.putInt(ARG_COLUMN_COUNT, columnCount);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        if (getArguments() != null) {
            mColumnCount = getArguments().getInt(ARG_COLUMN_COUNT);
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_filmroll_list, container, false);
        TrisquelDao dao = new TrisquelDao(this.getContext());
        dao.connection();
        list = dao.getAllFilmRolls();
        dao.close();

        // Set the adapter
        if (view instanceof RecyclerViewEmptySupport) {
            Context context = view.getContext();
            mRecyclerView = (RecyclerViewEmptySupport) view;
            mRecyclerView.setEmptyMessage(getString(R.string.warning_filmroll_not_registered));
            mRecyclerView.setEmptyView(container.findViewById(R.id.empty_view));
            if (mColumnCount <= 1) {
                mRecyclerView.setLayoutManager(new LinearLayoutManager(context));
            } else {
                mRecyclerView.setLayoutManager(new GridLayoutManager(context, mColumnCount));
            }

            filmrollRecyclerViewAdapter = new MyFilmRollRecyclerViewAdapter(list, mListener);
        }
        return view;
    }

    @Override
    public void onActivityCreated(final Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        mRecyclerView.setAdapter(filmrollRecyclerViewAdapter);
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof OnListFragmentInteractionListener) {
            mListener = (OnListFragmentInteractionListener) context;
        } else {
            throw new RuntimeException(context.toString()
                    + " must implement OnListFragmentInteractionListener");
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
    }

    public void insertFilmRoll(FilmRoll filmroll) {
        if (list != null) {
            int index = list.indexOf(filmroll);
            if (-1 == index) {
                list.add(0, filmroll);
                TrisquelDao dao = new TrisquelDao(this.getContext());
                dao.connection();
                long id = dao.addFilmRoll(filmroll);
                filmroll.id = (int)id;
                dao.close();
                Log.d("FilmRollFragment", "notifyItemInserted");
                filmrollRecyclerViewAdapter.notifyItemInserted(0);
            }
        }
    }

    public void updateFilmRoll(FilmRoll filmroll) {
        if (list != null){
            for(int i = 0; i < list.size(); i++){
                FilmRoll c = list.get(i);
                if(list.get(i).id == filmroll.id){
                    list.remove(i);
                    list.add(i, filmroll);
                    TrisquelDao dao = new TrisquelDao(this.getContext());
                    dao.connection();
                    dao.updateFilmRoll(filmroll);
                    ArrayList<Photo> p = dao.getPhotosByFilmRollId(filmroll.id);
                    filmroll.photos = p;
                    dao.close();
                    filmrollRecyclerViewAdapter.notifyItemChanged(i);
                }
            }
        }
    }

    public void refreshFilmRoll(int id) {
        if (list != null){
            for(int i = 0; i < list.size(); i++){
                FilmRoll c = list.get(i);
                if(list.get(i).id == id){
                    list.remove(i);
                    TrisquelDao dao = new TrisquelDao(this.getContext());
                    dao.connection();
                    FilmRoll f = dao.getFilmRoll(id);
                    ArrayList<Photo> p = dao.getPhotosByFilmRollId(id);
                    f.photos = p;
                    dao.close();
                    list.add(i, f);
                    Log.d("refreshFilmRoll", Integer.toString(id));
                    filmrollRecyclerViewAdapter.notifyItemChanged(i);
                }
            }
        }
    }

    public void deleteFilmRoll(int id){
        if (list != null){
            for(int i = 0; i < list.size(); i++){
                FilmRoll c = list.get(i);
                if(list.get(i).id == id){
                    list.remove(i);
                    TrisquelDao dao = new TrisquelDao(this.getContext());
                    dao.connection();
                    dao.deleteFilmRoll(id);
                    dao.close();
                    Log.d("deleteFilmRoll", Integer.toString(id));
                    filmrollRecyclerViewAdapter.notifyItemRemoved(i);
                }
            }
        }
    }

    /**
     * This interface must be implemented by activities that contain this
     * fragment to allow an interaction in this fragment to be communicated
     * to the activity and potentially other fragments contained in that
     * activity.
     * <p/>
     * See the Android Training lesson <a href=
     * "http://developer.android.com/training/basics/fragments/communicating.html"
     * >Communicating with Other Fragments</a> for more information.
     */
    public interface OnListFragmentInteractionListener {
        // TODO: Update argument type and name
        void onListFragmentInteraction(FilmRoll item, boolean isLong);
    }
}

File: app/src/main/java/net/tnose/app/trisquel/FilmRollFragment.kt
package net.tnose.app.trisquel

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.compose.foundation.ExperimentalFoundationApi
import androidx.compose.foundation.combinedClickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.livedata.observeAsState
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.ComposeView
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.unit.dp
import androidx.lifecycle.ViewModelProvider
import androidx.preference.PreferenceManager
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.Date
import java.util.TimeZone

class FilmRollFragment : androidx.fragment.app.Fragment() {
    private var mListener: OnListFragmentInteractionListener? = null
    private var mFilmRollViewModel: FilmRollViewModel? = null
    private var _currentFilter: Pair<Int, ArrayList<String>> = Pair(0, arrayListOf(""))

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val filtertype = arguments?.getInt(ARG_FILTER_TYPE) ?: 0
        val filterstr = arguments?.getStringArrayList(ARG_FILTER_VALUE) ?: arrayListOf("")
        _currentFilter = Pair(filtertype, filterstr)
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        mFilmRollViewModel = ViewModelProvider(this).get(FilmRollViewModel::class.java)

        return ComposeView(requireContext()).apply {
            setContent {
                MaterialTheme {
                    val filmrolls by mFilmRollViewModel!!.allFilmRollAndRels.observeAsState(initial = emptyList())
                    
                    FilmRollListScreen(
                        filmrolls = filmrolls,
                        onItemClick = { mListener?.onListFragmentInteraction(FilmRoll.fromEntity(it), false) },
                        onItemLongClick = { mListener?.onListFragmentInteraction(FilmRoll.fromEntity(it), true) },
                        emptyMessage = getString(R.string.warning_filmroll_not_registered)
                    )
                }
            }
        }
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        if (context is OnListFragmentInteractionListener) {
            mListener = context
        } else {
            throw RuntimeException(context.toString() + " must implement OnListFragmentInteractionListener")
        }
    }

    override fun onDetach() {
        super.onDetach()
        mListener = null
    }

    var currentFilter: Pair<Int, ArrayList<String>>
        get() = _currentFilter
        set(value) {
            _currentFilter = value
            val searchStr = when(value.first){
                1 -> value.second[0].toInt().toString()
                2 -> value.second[1]
                else -> ""
            }

            val pref = PreferenceManager.getDefaultSharedPreferences(this.requireContext())
            val key = pref.getInt("filmroll_sortkey", 0)
            mFilmRollViewModel!!.viewRule.value = Pair(key, Pair(value.first, searchStr))
        }

    fun changeSortKey(key: Int) {
        currentFilter = _currentFilter
    }

    fun insertFilmRoll(filmRoll: FilmRoll) {
        mFilmRollViewModel!!.insert(filmRoll.toEntity())
    }

    fun updateFilmRoll(filmRoll: FilmRoll) {
        mFilmRollViewModel!!.update(filmRoll.toEntity())
    }

    fun deleteFilmRoll(id: Int) {
        mFilmRollViewModel!!.delete(id)
    }

    interface OnListFragmentInteractionListener {
        fun onListFragmentInteraction(item: FilmRoll, isLong: Boolean)
    }

    companion object {
        private val ARG_FILTER_TYPE = "filter-type"
        private val ARG_FILTER_VALUE = "filter-val"

        fun newInstance(filterType: Int, filterValues: ArrayList<String>): FilmRollFragment {
            val fragment = FilmRollFragment()
            val args = Bundle()
            args.putInt(ARG_FILTER_TYPE, filterType)
            args.putStringArrayList(ARG_FILTER_VALUE, filterValues)
            fragment.arguments = args
            return fragment
        }
    }
}

fun getDateRange(dates: List<String>): String {
    if (dates.isEmpty()) return ""

    var minDate = Date(Long.MAX_VALUE)
    var maxDate = Date(0)
    val sdf = SimpleDateFormat("yyyy/MM/dd")
    sdf.timeZone = TimeZone.getTimeZone("UTC")

    for (date in dates) {
        var d = Date(0)
        try {
            d = sdf.parse(date) ?: Date(0)
        } catch (e: ParseException) {
        }

        if (minDate.after(d)) minDate = d
        if (maxDate.before(d)) maxDate = d
    }

    return if (minDate == maxDate) {
        sdf.format(minDate)
    } else {
        sdf.format(minDate) + "-" + sdf.format(maxDate)
    }
}

@OptIn(ExperimentalFoundationApi::class)
@Composable
fun FilmRollListScreen(
    filmrolls: List<FilmRollAndRels>,
    onItemClick: (FilmRollAndRels) -> Unit,
    onItemLongClick: (FilmRollAndRels) -> Unit,
    emptyMessage: String
) {
    if (filmrolls.isEmpty()) {
        Box(
            modifier = Modifier.fillMaxSize(),
            contentAlignment = Alignment.Center
        ) {
            Text(text = emptyMessage, style = MaterialTheme.typography.bodyLarge)
        }
    } else {
        LazyColumn(
            modifier = Modifier.fillMaxSize()
        ) {
            items(filmrolls, key = { it.filmRoll.id }) { item ->
                Column(
                    modifier = Modifier
                        .fillMaxWidth()
                        .combinedClickable(
                            onClick = { onItemClick(item) },
                            onLongClick = { onItemLongClick(item) }
                        )
                        .padding(horizontal = 16.dp, vertical = 16.dp)
                ) {
                    if (item.filmRoll.name.isEmpty()) {
                        Text(
                            text = stringResource(id = R.string.empty_name),
                            style = MaterialTheme.typography.titleLarge,
                            fontStyle = FontStyle.Italic
                        )
                    } else {
                        Text(
                            text = item.filmRoll.name,
                            style = MaterialTheme.typography.titleLarge
                        )
                    }
                    
                    val cameraBrandStr = "${item.camera?.manufacturer ?: ""} ${item.camera?.modelName ?: ""}   ${item.filmRoll.manufacturer} ${item.filmRoll.brand}"
                    Text(
                        text = cameraBrandStr,
                        style = MaterialTheme.typography.bodyMedium,
                        color = MaterialTheme.colorScheme.onSurfaceVariant
                    )

                    val exp = item.photoDates.size
                    val array = arrayListOf<String>()
                    val dateRange = getDateRange(item.photoDates)
                    if (dateRange.isNotEmpty()) array.add(dateRange)
                    array.add(if (exp == 1) "%d shot".format(exp) else "%d shots".format(exp))
                    
                    Text(
                        text = array.joinToString("   "),
                        style = MaterialTheme.typography.bodySmall,
                        color = MaterialTheme.colorScheme.onSurfaceVariant
                    )
                }
            }
        }
    }
}
